### PR TITLE
Use ufcs in derive(Debug)

### PR DESCRIPTION
**Pull request by pnkfelix**
Cc #81211.

(Arguably this *is* the fix for it.)

### CODE REVIEW

**Comment by pnkfelix** [comment]
There is one thing I was curious about: I used `cx.std_path` because I thought that was necessary to get the right behavior with respect to whether `core` or `std` is emitted, but I do see use of some sort of `path_std!` macro, which seems much nicer.

Should I be using `path_std!` here as well, then? I plan to give it a whirl at some point. But I figured I could put this up for initial review feedback in the meantime.

**Comment by pnkfelix** [comment]
(I'm seeing that failure locally too, but I'm at a loss for how what I did could have caused it...)

**Comment by pnkfelix** [comment]
(marking as S-waiting-on-author since it cannot land as is, but I still would welcome any hints as to what I might have done wrong.)

**Comment by oli-obk** [comment]
I believe I have seen the opposite change recently. This may be a spuriously fluctuating diagnostic (or semi-spurious, depending on random but fixed things in the compiler, since I have not seen it happen randomly on CI).

File: compiler/rustc_builtin_macros/src/deriving/debug.rs
Comment (oli-obk) [inline compiler/rustc_builtin_macros/src/deriving/debug.rs:82]: Maybe we should just remove the `expr_method_call` function entirely (or perma-deprecate it so ppl searching for it get an explanation of why they should use ufcs)

**Review by oli-obk — APPROVED** [review]
r=me with the "spurious" test blessed. I'll open an issue about it

File: compiler/rustc_builtin_macros/src/deriving/debug.rs
Comment (pnkfelix) [reply compiler/rustc_builtin_macros/src/deriving/debug.rs:82]: Yeah I considered this as well. Probably a good idea, especially if this is the last use of it. I'll try to look into that.

**Comment by pnkfelix** [comment]
Hmm, do I need to wait for PR #81393 to land before I can land this, then?

**Comment by oli-obk** [comment]
I guess? Let's see

@bors r+

**Comment by pnkfelix** [comment]
oh you've got to be kidding me, my local systems failure to test coverage is biting me *again*, just like in PR #81257?

Okay, I guess its time to invest a little more effort in figuring out what is supposed to be running and why my system isn't running it...

**Comment by tmiasko** [comment]
Perf results: https://perf.rust-lang.org/compare.html?start=6ad11e2e25919b75ebbc36d7910f2a1126a7e873&end=186f7ae5b04d31d8ccd1746ac63cdf1ab4bc2354&stat=instructions:u

**Comment by tmiasko** [comment]
The experiment from #81760, while produces smaller MIR, didn't help with regressions. The comparisons before & #81760 (insert extra caveats about the non-adjacent comparison ...) https://perf.rust-lang.org/compare.html?start=6ad11e2e25919b75ebbc36d7910f2a1126a7e873&end=d589fc7d554c7ecdab26eb7ae07fd6dc7e8280f7&stat=instructions%3Au

There are additional queries for `type_op_ascribe_user_type` and regression in typeck. The new MIR does indeed contain additinal user type annotations. Looking at how they are used, they will involve extra work equating & relating the types, so this might be an aspect that contributes to the regressions.

**Comment by oli-obk** [comment]
Oh... so UFCs always were less efficient than method calls in terms of compile-time or at least query invocation count. I wonder if we can adjust the relevant call sites to not use `type_op_ascribe_user_type` where it's "obviously" not necessary. Potentially by wrapping that query in a function that does some early aborts.

Not sure what changed in typeck, will need to dig into the code in order to figure that out

**Pull request merged**
